src/binary-reader.cc

```diff
@@ -1557,12 +1558,13 @@ Result BinaryReader::ReadInstructions(bool stop_on_end,
       case Opcode::I32AtomicStore:
       case Opcode::I64AtomicStore: {
         Address alignment_log2;
-        CHECK_RESULT(ReadAlignment(&alignment_log2, "store alignment"));
+        Index memidx;
+        CHECK_RESULT(ReadAlignMem(&alignment_log2, &memidx, "store alignment"));
         Address offset;
         CHECK_RESULT(ReadAddress(&offset, 0, "store offset"));
 
-        CALLBACK(OnAtomicStoreExpr, opcode, alignment_log2, offset);
-        CALLBACK(OnOpcodeUint32Uint32, alignment_log2, offset);
+        CALLBACK(OnAtomicStoreExpr, opcode, memidx, alignment_log2, offset);
+        CHECK_RESULT(CallbackAlignMemOffset(memidx, alignment_log2, offset));
```

Comment:
Nice!  This change certainly looks like a step in the right direction.

What do you think about combining `ReadAlignMem` with `ReadAddress` into a single function `ReadMemroyLocation` which would replace `ReadAlignMem` and include the reading of the offset?  Could that function also include `CallbackAlignMemOffset` (removing the need for that new function)?

Reply:
Done. I couldn't quite combine it with the CallbackAlignMemOffset because that needs to happen after the callback for the actual expression (otherwise all the opcnts and dumps are off), but I was able to clean it up a little.